src/hotspot/share/opto/vector.cpp

```diff
@@ -354,42 +354,43 @@ Node* PhaseVector::expand_vbox_node_helper(Node* vec_box,
     return new_phi;
   } else if (vbox->is_Proj() && vbox->in(0)->Opcode() == Op_VectorBoxAllocate) {
     VectorBoxAllocateNode* vbox_alloc = static_cast<VectorBoxAllocateNode*>(vbox->in(0));
-    return expand_vbox_alloc_node(vec_box, vbox_alloc, vect, box_type, vect_type);
+    return expand_vbox_alloc_node(vbox_alloc, vect, box_type, vect_type);
   } else {
     assert(!vbox->is_Phi(), "");
     // TODO: assert that expanded vbox is initialized with the same value (vect).
     return vbox; // already expanded
   }
 }
 
-Node* PhaseVector::expand_vbox_alloc_node_mf(Node* vbox,
-                                             VectorBoxAllocateNode* vbox_alloc,
-                                             Node* value,
-                                             const TypeInstPtr* box_type,
-                                             const TypeVect* vect_type) {
-  assert(vbox->isa_InlineType(), "");
+Node* PhaseVector::expand_vbox_alloc_node_vector(VectorBoxAllocateNode* vbox_alloc,
+                                                 Node* value,
+                                                 const TypeInstPtr* box_type,
+                                                 const TypeVect* vect_type) {
   JVMState* jvms = clone_jvms(C, vbox_alloc);
   GraphKit kit(jvms);
   PhaseGVN& gvn = kit.gvn();
 
-  ciInstanceKlass* box_klass = box_type->instance_klass();
+  ciInlineKlass* box_klass = static_cast<ciInlineKlass*>(box_type->inline_klass());
+
   BasicType bt = vect_type->element_basic_type();
   int num_elem = vect_type->length();
   int elem_size = type2aelembytes(bt);
 
   const TypeKlassPtr* klass_type = box_type->as_klass_type();
   Node* klass_node = kit.makecon(klass_type);
-  Node* buffer_mem = kit.new_instance(klass_node, NULL, NULL, /* deoptimize_on_exception */ true, vbox->as_InlineType());
-  // TODO: Re-use existing value storage routine from InlineTypeNode.
-  //vbox->as_InlineType()->store(&kit, buffer, buffer, box_klass);
-
-  // Store the vector value into the array.
-  // (The store should be captured by InitializeNode and turned into initialized store later.)
-  ciSymbol* payload_sig = ciSymbol::make(VectorSupport::get_vector_payload_field_signature(bt, num_elem)->as_C_string());
-  ciSymbol* payload_name = ciSymbol::make(vmSymbols::payload_name()->as_C_string());
-  ciField* payload = box_klass->get_field_by_name(payload_name, payload_sig, false);
+  Node* buffer_mem = kit.new_instance(klass_node, NULL, NULL, /* deoptimize_on_exception */ true);
+
+  assert(VectorSupport::is_vector(box_klass), "");
+  ciField* payload_field = box_klass->declared_nonstatic_field_at(0);
+  int offset = payload_field->offset();
+  if (!payload_field->is_flattened()) {
+    ciInlineKlass* payload_klass = static_cast<ciInlineKlass*>(payload_field->type());
+    assert(VectorSupport::is_vector_payload_mf(payload_klass), "");
+    ciField* mutifield = payload_klass->declared_nonstatic_field_at(0);
+    offset += mutifield->offset();
+  }
```

Comment:
If the `payload_field` is not flattened, I'm afraid the payload value should be buffered like the original array instance?

Reply:
Payloads are primitive type objects of VectorPayloadMF* classes, by default should be flattened in the containing object i.e. container object size will already account for flattened field size, but flattening happens under the influence of InlineFieldMaxFlatSize runtime flag, above code handles non-flattened case.

Reply:
Yes, the non-flattened case may exist. What I mean is when handling the non-flattened payloads, should a buffer be allocated for it as a instance like before?